pkg/controllers/tlspolicy/tlspolicy_controller.go

```diff
@@ -0,0 +1,231 @@
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// add finalizer to the tlsPolicy
+	if !controllerutil.ContainsFinalizer(tlsPolicy, TLSPolicyFinalizer) {
+		if err := r.AddFinalizer(ctx, tlsPolicy, TLSPolicyFinalizer); client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{Requeue: true}, err
+		} else if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+	}
+
+	specErr := r.reconcileResources(ctx, tlsPolicy, targetNetworkObject)
+
+	newStatus := r.calculateStatus(tlsPolicy, specErr)
+	tlsPolicy.Status = *newStatus
+
+	if !equality.Semantic.DeepEqual(previous.Status, tlsPolicy.Status) {
+		updateErr := r.Client().Status().Update(ctx, tlsPolicy)
+		if updateErr != nil {
+			// Ignore conflicts, resource might just be outdated.
+			if apierrors.IsConflict(updateErr) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+			return ctrl.Result{}, updateErr
+		}
+	}
+
+	if specErr != nil {
+		return ctrl.Result{}, specErr
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *TLSPolicyReconciler) reconcileResources(ctx context.Context, tlsPolicy *v1alpha1.TLSPolicy, targetNetworkObject client.Object) error {
+	// validate
+	err := tlsPolicy.Validate()
+	if err != nil {
+		return err
+	}
+
+	// reconcile based on gateway diffs
+	gatewayDiffObj, err := r.ComputeGatewayDiffs(ctx, tlsPolicy, targetNetworkObject, &TLSPolicyRefsConfig{})
+	if err != nil {
+		return err
+	}
+
+	if err := r.reconcileCertificates(ctx, tlsPolicy, gatewayDiffObj); err != nil {
+		return err
+	}
+
+	// set direct back ref - i.e. claim the target network object as taken asap
+	if err := r.ReconcileTargetBackReference(ctx, client.ObjectKeyFromObject(tlsPolicy), targetNetworkObject, TLSPolicyBackRefAnnotation); err != nil {
+		return err
+	}
+
+	// set annotation of policies affecting the gateway - should be the last step, only when all the reconciliation steps succeed
+	return r.ReconcileGatewayPolicyReferences(ctx, tlsPolicy, gatewayDiffObj)
+}
+
+func (r *TLSPolicyReconciler) deleteResources(ctx context.Context, tlsPolicy *v1alpha1.TLSPolicy, targetNetworkObject client.Object) error {
+	// delete based on gateway diffs
+
+	gatewayDiffObj, err := r.ComputeGatewayDiffs(ctx, tlsPolicy, targetNetworkObject, &TLSPolicyRefsConfig{})
+	if err != nil {
+		return err
+	}
+
+	if err := r.reconcileCertificates(ctx, tlsPolicy, gatewayDiffObj); err != nil {
+		return err
+	}
+
+	// remove direct back ref
+	if targetNetworkObject != nil {
+		if err := r.DeleteTargetBackReference(ctx, client.ObjectKeyFromObject(tlsPolicy), targetNetworkObject, TLSPolicyBackRefAnnotation); err != nil {
+			return err
+		}
+	}
+
+	// update annotation of policies affecting the gateway
+	return r.ReconcileGatewayPolicyReferences(ctx, tlsPolicy, gatewayDiffObj)
+}
+
+func (r *TLSPolicyReconciler) calculateStatus(tlsPolicy *v1alpha1.TLSPolicy, specErr error) *v1alpha1.TLSPolicyStatus {
+	newStatus := tlsPolicy.Status.DeepCopy()
```

Comment:
Some thoughts on TLSPolicy status in https://github.com/Kuadrant/multicluster-gateway-controller/issues/374#issuecomment-1650172511 if you haven't already seen those comments.

Reply:
Yeah at least a ready condition.

Reply:
There is a ready condition that will be set to false on spec reconcile failure, with a reason of `ReconciliationError` and message containing the error. I guess if we want more tailored messages based on types of errors we can look into that, but this is a bit of a catch all currently.

Reply:
I think we should try and keep ourselves aligned with other kuadrant operators and look to do whatever ever is agreed here https://github.com/Kuadrant/architecture/pull/9